Revamp keras-mxnet docs

sandeep-krishnamurthy opened the pull request:
Revamping Keras-MXNet documentation.

Following are still pending:
1. Tutorial for save_mxnet_model API
2. Clean up benchmark/ docs.

@roywei @kalyc @aaronmarkham

kalyc commented on docs/mxnet_backend/installation.md:
```diff
@@ -135,30 +138,25 @@ $  nvidia-smi
-`Option 2` Edit the keras config file
-
-Open "*~/.keras/keras.json*" file and set "*backend=mxnet*".
+We strongly recommend changing the image_data_format to `channels_first`. MXNet is significantly faster on 'channels_first' data. Default is set to 'channels_last' with an objective to be compatible with majority of existing users of Keras. See [performance tips guide](performance_guide.md) for more details.
```
are we showing this message prior to model execution as well in the cli?

sandeep-krishnamurthy replied:
We show this message when user loads keras with mxnet backend.

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
````diff
@@ -0,0 +1,106 @@
+image_data_format: channels_last
+```
+
+We strongly recommend changing the image_data_format to `channels_first`. MXNet is significantly faster on 'channels_first' data. 'channels_first' data format is optimal for training on NVIDIA GPUs with cuDNN, see [here](https://www.tensorflow.org/performance/performance_guide#data_formats) for more details.
````
don't link "here"...

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
````diff
@@ -0,0 +1,106 @@
+
+```
+NOTE:
+    Keras dataset API is not consistent across the different dataset. It returns few datasets based on the 
````
across different datasets. It returns a few datasets...

...and a few datasets in the channels...

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
````diff
@@ -0,0 +1,106 @@
+    MNIST dataset). It is recommended to always verify the shape of data returned.
+```
+
+## Step 3. Update input_shape argument when constructing the model
````
Update the...

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+## Step 3. Update input_shape argument when constructing the model
+
+You need to update the input shapes in the network to channels first format. For example, using Conv2D on the color images with size 256x256, you need to set:
```
256.

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
````diff
@@ -0,0 +1,106 @@
+    model.add(Conv2D(32, (3, 3), padding='same', input_shape=(256,256,3)))
+```
+
+For channels first format:
````
...format, set the `input_shape` parameter as follows:

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+You need to update the input shapes in the network to channels first format. For example, using Conv2D on the color images with size 256x256, you need to set:
+
+For channels last format:
```
...format, set the `input_shape` parameter as follows:

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+# Optimized MXNet binaries
+
+MXNet PyPi packages are distributed in multiple flavors each optimized for a specific underlying hardware. Please install the correct MXNet flavor optimized for your instance type for the best performance.
```
flavors, where each...

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+4. Optimized for NVIDIA GPUs and Intel CPUs: `pip install mxnet-cu90-mkl`
+
+5. Refer installation guide for more configurations to build from source: [MXNet Installation guide](https://mxnet.incubator.apache.org/install/index.html) 
```
Refer to the [MXNet installation guide]()...

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+5. Refer installation guide for more configurations to build from source: [MXNet Installation guide](https://mxnet.incubator.apache.org/install/index.html) 
+
+Please see [MXNet Performance Guide](https://mxnet.incubator.apache.org/faq/perf.html) for more detailed performance tips for MXNet.
```
see the MXNet...
delete "for MXNet"

aaronmarkham commented on docs/mxnet_backend/performance_guide.md:
```diff
@@ -0,0 +1,106 @@
+
+# Using MXNet Profiler
+
+MXNet has a built-in profiler that provides detailed information about the execution times for each operator. MXNet profiler complements general profiling tools like `nvprof` and `gprof` by summarizing execution time at the operator level, rather than at a function, kernel, or instruction level. Follow [MXNet Performance Guide](https://mxnet.incubator.apache.org/faq/perf.html) for more details on ...
```
...operator. The MXNet profiler...
Follow --> Review the MXNet...
delete "on MXNet profiler"

aaronmarkham commented on docs/mxnet_backend/using_rnn_with_mxnet_backend.md:
```diff
@@ -0,0 +1,69 @@
+
+# Overview
+
+In this document, we describe the limitations of using RNNs with MXNet backend and workarounds for the same.
```
and available workarounds.

aaronmarkham commented on docs/mxnet_backend/using_rnn_with_mxnet_backend.md:
```diff
@@ -0,0 +1,69 @@
+
+# Variable length inputs are not supported
+
+MXNet backend does not support variable length inputs in the Recurrent Layers. To overcome this limitation, you can pad the input sequences to prepare fixed length inputs. MXNet backend requires `input_shape` and `unroll=True` parameters while adding SimpleRNN/LSTM/GRU layer.
```
recurrent layers.
The MXNet backend requires both the... and the... adding the

aaronmarkham commented on docs/mxnet_backend/using_rnn_with_mxnet_backend.md:
````diff
@@ -0,0 +1,69 @@
+
+```
+NOTE:
+    MXNet does not support symbolic control flow operators. However, this is a work in progress feature. This feature will be soon supported in upcoming releases.
````
...be supported in an upcoming release.

aaronmarkham commented on docs/mxnet_backend/using_rnn_with_mxnet_backend.md:
```diff
@@ -0,0 +1,69 @@
+
+You should pad the variable length input sequences to make it a fixed length. You can use Keras API - `keras.preprocessing.sequence.pad_sequences` for padding the input.
+ 
+Below we provide an example usecase for the same:
```
Below is an example use case:

aaronmarkham commented on docs/mxnet_backend/using_rnn_with_mxnet_backend.md:
````diff
@@ -0,0 +1,69 @@
+```
+# Unroll=False is not supported
+
+As described above, MXNet backend does not support variable length input and hence `unroll=False` is not supported in RNN layers. Users are expected to provide `input_shape` and set `unroll=True`.
````
Users --> You

sandeep-krishnamurthy commented:
@aaronmarkham - Thanks for details review comments. I have addressed all your review comments.

Added another tutorial - docs/mxnet_backend/save_mxnet_model.md

sandeep-krishnamurthy commented on docs/mxnet_backend/save_mxnet_model.md:
```diff
@@ -0,0 +1,175 @@
+# ... Assuming you have built and trained a Model
+model.fit(x_train, y_train,
+          batch_size=batch_size,
+          epochs=epochs,
```
Did not understand this.

sandeep-krishnamurthy commented on docs/mxnet_backend/save_mxnet_model.md:
```diff
@@ -0,0 +1,175 @@
+# Step1: Load the model in MXNet
+
+# Use the same prefix and epoch parameters we used in save_mxnet_model API.
+sym, arg_params, aux_params = mx.model.load_checkpoint(prefix='mnist_cnn', epoch=0)
```
No, it is a required parameter for mx.model.load_checkpoint() API

aaronmarkham replied:
But you're using epoch=0, when we just did training in the previous section. Should it not match the amount of training done? Or should it say that epoch doesn't really matter so you can just pass 0?

sandeep-krishnamurthy commented on docs/mxnet_backend/save_mxnet_model.md:
````diff
@@ -0,0 +1,175 @@
+print("Actual - ", labels[idx])
+```
+
+That's it! We trained a CNN model with Keras interface and used MXNet native engine in Python for inference. Also 
````
Very nice suggestion, we will try to have a new tutorial for using a pre-trained Keras model(h5py file) for ResNet56-CIFAR10 and export as mxnet model, perform inference in MXNet. Will track this in this Github issue post-release - https://github.com/awslabs/keras-apache-mxnet/issues/84

sandeep-krishnamurthy commented:
Thanks @aaronmarkham for the feedback. I have addressed your new set of comments.

Pending:
Revamp benchmark results documentation.

sandeep-krishnamurthy commented:
@aaronmarkham  - Added the final document for your review - benchmark/README.md

sandeep-krishnamurthy commented:
This PR is now complete with all docs for mxnet backend rewritten/revamped. 
Pending:
1. Review from @aaronmarkham for benchmark/README.md file.
2. Updating latest benchmark results in benchmark/README.md by @roywei 

@lupesko - Any other comments?

aaronmarkham commented on benchmark/benchmark_result/RNN_result.md:
```diff
@@ -1,3 +1,40 @@
+
 # Detailed RNN Benchmark Results
 
 Below is the result of GPU memory usage while running LSTM model on Synthetic, Nietzsche, and WikiText-2 character level dataset.
```
This is left for the reader's interpretation, and I'm not sure what to think about it. No commentary to assist?
